sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp

```diff
@@ -237,13 +238,14 @@ __SYCL_TYPE(annotated_ptr) annotated_ptr<T, detail::properties_t<Props...>> {
   annotated_ptr(const annotated_ptr &) = default;
   annotated_ptr &operator=(const annotated_ptr &) = default;
 
+  static_assert(std::is_same_v<T, void> || std::is_trivially_copyable_v<T>,
```

Comment:
why do you allow void here? If we do the spec should reflect that. But how is this useful? You read from it because the implicit conversion from ann_ref doesn't work.

Reply:
It seems void is a non-trivially-copyable type, which surprises me a bit. Thus I added an expectation here.

I guess I should update the spec as well?

Reply:
Even though a ann_ptr<void> can't be dereferenced it is valid to call get on it. This might be useful in some cases to have a ann_ptr which can't be dereferenced but creating it and unwrapping it (calling get) works. If we want to support that we should support that for all non-trivial-copyable types and instead of saying that ann_ptr can't be instantiated for non-trivial-copyable types that ann_ref can't be instantiated for those types. And therefore you can create such a ann_ptr but you can't deference it.

Reply:
The spec currently says that the annotations have an effect when the `annotated_ref` is dereferenced, and it does not state any other time that they are guaranteed to have an effect.  If we don't allow creation of the `annotated_ref`, then there is no guarantee that the annotations have any effect.  Therefore, I think we should say that `T` must be trivially copyable, and this means that `void` is prohibited.

Reply:
> If we don't allow creation of the annotated_ref, then there is no guarantee that the annotations have any effect. 

Correct. But if a user is using `get` for a trivially copyable type there is no guarantee either. And I suspect we agree we don't want to remove `get` because we want to give users the freedom to use annotated_ptr in this way.

I can think of multiple use-cases of wanting an annotated_ptr which can't be dereferenced (prior to unwrapping with get):
1) a user might need a polymorphic `void*` pointer (the type can change at runtime). They would pass a ann_ptr<void> and would cast it to an ann_ptr<T>, where T is the actual type and trivially-copyable before dereferencing it. Note that this is one typical use-case for `void*` and that a raw void* pointer can't be dereferenced either.
2) the user might not care about the annotation being applied on load/store rather they want the property to do static_dispatch on it (e.g. have a fast or slow algorithm depending on whether the pointer is aligned).
3) the user might want to apply the annotation manually when loading/storing and therefore only need the ann_ptr at the interface but don't care about ann_ref.

Reply:
We agreed here that `void` is allowed in annotated_ptr as an exception. We forbid non_trivially_copyable type in both `annotated_ptr` and `annotated_ref` for now and we shall see if we need to enable `annotated_ptr` with non_trivially_copyable  in the future